doc/message-structure.md

```diff
@@ -0,0 +1,20 @@
+# Log message structure
+
+Within monolog log messages are passed around as arrays, for example to processors or handlers.
+The table below describes which keys are always available for every log message.
+
+key        | type                      | description
+-----------|---------------------------|-------------------------------------------------------------------------------
+message    | string                    | The log message. This string may contain placeholders that will be replaced by variables from the context, e.g., "User %username% logged in" with `['username' => 'John']` as context will be written as "User John logged in".
+level      | int                       | Severity of the log message. See log levels described in [01-usage.md](01-usage.md).
+level_name | string                    | String representation of log level.
+context    | array                     | Arbirtrary data passed with the constuction of the message. For example the username of the current user or their IP addres.
+channel    | string                    | The channel this message was logged to.
+datetime   | Monolog\DateTimeImmutable | Date and time when the message was logged. Class extends `\DateTimeImmutable`.
+extra      | array                     | A placeholder array where processors can put additional data. Always available, but empty if there are no processors registered.
+
+Some handlers and/or processors may add additional keys to the array.
```

Comment:
Processors should not add new keys. This is not a supported usage pattern. They should store them in ``extra``.

And handlers are not adding keys to the record either (they could be turning the record into another data structure needed by the target system though of course)